File: edu.wpi.first.wpilib.plugins.cpp/resources/templates/examples/UltrasonicPID/src/Robot.cpp
#include <AnalogInput.h>
#include <IterativeRobot.h>
#include <PIDController.h>
#include <PIDOutput.h>
#include <RobotDrive.h>

/**
 * This is a sample program demonstrating how to use an ultrasonic sensor and
 * proportional control to maintain a set distance from an object.
 */
class Robot: public IterativeRobot {
public:
	/**
	 * Drives the robot a set distance from an object using PID control and the
	 * ultrasonic sensor.
	 */
	void TeleopInit() {
		// Set expected range to 0-24 inches; e.g. at 24 inches from object go
		// full forward, at 0 inches from object go full backward.
		pidController.SetInputRange(0, 24 * kValueToInches);
		// Set setpoint of the pidController
		pidController.SetSetpoint(kHoldDistance * kValueToInches);
		pidController.Enable(); // begin PID control
	}

private:
	// internal class to write to myRobot (a RobotDrive object) using a PIDOutput
	class MyPIDOutput: public PIDOutput {
	public:
		MyPIDOutput(RobotDrive& r) :
				rd(r) {
			rd.SetSafetyEnabled(false);
		}

		void PIDWrite(double output) {
			rd.Drive(output, 0); // write to myRobot (RobotDrive) by reference
		}
	private:
		RobotDrive& rd;
	};

	// Distance in inches the robot wants to stay from an object
	static constexpr int kHoldDistance = 12;

	// Factor to convert sensor values to a distance in inches
	static constexpr double kValueToInches = 0.125;

	AnalogInput ultrasonic { 0 };
	RobotDrive myRobot { 0, 1 };
	PIDController pidController { 7, 0.018, 1.5, &ultrasonic,
			new MyPIDOutput(myRobot) };
};

START_ROBOT_CLASS(Robot)

File: edu.wpi.first.wpilib.plugins.java/resources/templates/examples/UltrasonicPID/src/org/usfirst/frc/team190/robot/Robot.java
package $package;

import edu.wpi.first.wpilibj.AnalogInput;
import edu.wpi.first.wpilibj.IterativeRobot;
import edu.wpi.first.wpilibj.PIDController;
import edu.wpi.first.wpilibj.PIDOutput;
import edu.wpi.first.wpilibj.RobotDrive;

/**
 * This is a sample program to demonstrate the use of a PIDController with an
 * ultrasonic sensor to reach and maintain a set distance from an object.
 */
public class Robot extends IterativeRobot {
	// distance in inches the robot wants to stay from an object
	private static final double kHoldDistance = 12.0;

	// maximun distance in inches we expect the robot to see
	private static final double kMaxDistance = 24.0;

	// factor to convert sensor values to a distance in inches
	private static final double kValueToInches = 0.125;

	// proportional speed constant
	private static final double kP = 7.0;

	// integral speed constant
	private static final double kI = 0.018;

	// derivative speed constant
	private static final double kD = 1.5;

	private static final int kLeftMotorPort = 0;
	private static final int kRightMotorPort = 1;
	private static final int kUltrasonicPort = 0;

	private AnalogInput ultrasonic = new AnalogInput(kUltrasonicPort);
	private RobotDrive myRobot = new RobotDrive(kLeftMotorPort, kRightMotorPort);
	private PIDController pidController = new PIDController(kP, kI, kD, ultrasonic, new MyPidOutput());

	/**
	 * Drives the robot a set distance from an object using PID control and the
	 * ultrasonic sensor.
	 */
	@Override
	public void teleopInit() {
		// Set expected range to 0-24 inches; e.g. at 24 inches from object go
		// full forward, at 0 inches from object go full backward.
		pidController.setInputRange(0, kMaxDistance * kValueToInches);
		// Set setpoint of the pidController
		pidController.setSetpoint(kHoldDistance * kValueToInches);
		pidController.enable(); // begin PID control
	}

	private class MyPidOutput implements PIDOutput {
		@Override
		public void pidWrite(double output) {
			myRobot.drive(output, 0);
		}
	}
}
